File: scripts/ci/src/ci/lib/catalog.py
"""Shared target catalogs for CI pipeline generation.

Central registry of all build targets, deploy sites, Helm charts,
and deployment mappings used across CI scripts.
"""

from __future__ import annotations

from dataclasses import dataclass, field

# ---------------------------------------------------------------------------
# Container image push targets
# ---------------------------------------------------------------------------

IMAGE_PUSH_TARGETS: list[dict[str, str]] = [
    {"target": "//packages/birmel:image_push", "version_key": "shepherdjerred/birmel", "name": "birmel"},
    {"target": "//packages/sentinel:image_push", "version_key": "shepherdjerred/sentinel", "name": "sentinel"},
    {
        "target": "//packages/tasknotes-server:image_push",
        "version_key": "shepherdjerred/tasknotes-server",
        "name": "tasknotes-server",
    },
    {
        "target": "//packages/scout-for-lol:image_push",
        "version_key": "shepherdjerred/scout-for-lol/beta",
        "name": "scout-for-lol",
    },
    {
        "target": "//packages/discord-plays-pokemon:image_push",
        "version_key": "shepherdjerred/discord-plays-pokemon",
        "name": "discord-plays-pokemon",
    },
    {
        "target": "//packages/starlight-karma-bot:image_push",
        "version_key": "shepherdjerred/starlight-karma-bot/beta",
        "name": "starlight-karma-bot",
    },
    {
        "target": "//packages/better-skill-capped/fetcher:image_push",
        "version_key": "shepherdjerred/better-skill-capped-fetcher",
        "name": "better-skill-capped-fetcher",
    },
    {
        "target": "//tools/oci:obsidian_headless_push",
        "version_key": "shepherdjerred/obsidian-headless",
        "name": "obsidian-headless",
    },
    {
        "target": "//packages/status-page/api:image_push",
        "version_key": "shepherdjerred/status-page-api",
        "name": "status-page-api",
    },
]

INFRA_PUSH_TARGETS: list[dict[str, str]] = [
    {"target": "//packages/homelab/src/ha:image_push", "version_key": "shepherdjerred/homelab", "name": "homelab"},
    {
        "target": "//packages/homelab/src/deps-email:image_push",
        "version_key": "shepherdjerred/dependency-summary",
        "name": "dependency-summary",
    },
    {
        "target": "//packages/homelab/src/dns-audit:image_push",
        "version_key": "shepherdjerred/dns-audit",
        "name": "dns-audit",
    },
    {
        "target": "//packages/homelab/src/caddy-s3proxy:image_push",
        "version_key": "shepherdjerred/caddy-s3proxy",
        "name": "caddy-s3proxy",
    },
]

# ---------------------------------------------------------------------------
# npm packages
# ---------------------------------------------------------------------------

NPM_PACKAGES: list[dict[str, str]] = [
    {"name": "bun-decompile", "dir": "packages/bun-decompile"},
    {"name": "astro-opengraph-images", "dir": "packages/astro-opengraph-images"},
    {"name": "webring", "dir": "packages/webring"},
    {"name": "helm-types", "dir": "packages/homelab/src/helm-types"},
]

# ---------------------------------------------------------------------------
# Static site deploys
# ---------------------------------------------------------------------------

DEPLOY_SITES: list[dict[str, str | bool]] = [
    {
        "bucket": "sjer-red",
        "name": "sjer.red",
        "build_dir": "packages/sjer.red",
        "build_cmd": "bun run astro build",
        "dist_dir": "packages/sjer.red/dist",
        "needs_playwright": True,
        "workspace_deps": "astro-opengraph-images,webring",
    },
    {
        "bucket": "clauderon",
        "name": "clauderon docs",
        "build_dir": "packages/clauderon/docs",
        "build_cmd": "bun run astro build",
        "dist_dir": "packages/clauderon/docs/dist",
        "workspace_deps": "astro-opengraph-images",
    },
    {
        "bucket": "resume",
        "name": "resume",
        "build_dir": "packages/resume",
        "build_cmd": "",
        "dist_dir": "packages/resume",
    },
    {
        "bucket": "webring",
        "name": "webring",
        "build_dir": "packages/webring",
        "build_cmd": "bun run typedoc",
        "dist_dir": "packages/webring/docs",
    },
    {
        "bucket": "cook",
        "name": "cooklang-rich-preview",
        "build_dir": "packages/cooklang-rich-preview",
        "build_cmd": "bun run astro build",
        "dist_dir": "packages/cooklang-rich-preview/dist",
    },
    {
        "bucket": "status-page",
        "name": "status-page",
        "build_dir": "packages/status-page/web",
        "build_cmd": "bun run astro build",
        "dist_dir": "packages/status-page/web/dist",
    },
]

# ---------------------------------------------------------------------------
# OpenTofu stacks
# ---------------------------------------------------------------------------

TOFU_STACKS: list[str] = ["cloudflare", "github", "seaweedfs"]

TOFU_STACK_LABELS: dict[str, str] = {
    "cloudflare": "Cloudflare DNS",
    "github": "GitHub Config",
    "seaweedfs": "SeaweedFS Config",
}

# ---------------------------------------------------------------------------
# Package-to-site mapping (for change detection)
# ---------------------------------------------------------------------------

PACKAGE_TO_SITE: dict[str, str] = {
    "sjer.red": "sjer-red",
    "resume": "resume",
    "clauderon": "clauderon",
    "webring": "webring",
    "cooklang-rich-preview": "cook",
}

# ---------------------------------------------------------------------------
# Helm charts
# ---------------------------------------------------------------------------

HELM_CHARTS: list[str] = [
    "ddns",
    "apps",
    "scout-beta",
    "scout-prod",
    "starlight-karma-bot-beta",
    "starlight-karma-bot-prod",
    "redlib",
    "better-skill-capped-fetcher",
    "plausible",
    "birmel",
    "cloudflare-tunnel",
    "media",
    "home",
    "postal",
    "syncthing",
    "golink",
    "freshrss",
    "pokemon",
    "gickup",
    "grafana-db",
    "mcp-gateway",
    "s3-static-sites",
    "kyverno-policies",
    "bugsink",
    "dns-audit",
    "sentinel",
    "tasknotes",
    "bazel-remote",
    "status-page",
]

# ---------------------------------------------------------------------------
# Version keys for digest tracking
# ---------------------------------------------------------------------------

VERSION_KEYS: list[str] = [
    "shepherdjerred/homelab",
    "shepherdjerred/dependency-summary",
    "shepherdjerred/dns-audit",
    "shepherdjerred/caddy-s3proxy",
    "shepherdjerred/sentinel",
    "shepherdjerred/birmel",
    "shepherdjerred/tasknotes-server",
    "shepherdjerred/obsidian-headless",
    "shepherdjerred/starlight-karma-bot/beta",
    "shepherdjerred/better-skill-capped-fetcher",
    "shepherdjerred/discord-plays-pokemon",
    "shepherdjerred/scout-for-lol/beta",
    "shepherdjerred/status-page-api",
]

# ---------------------------------------------------------------------------
# Target aliases (user-friendly shorthand)
# ---------------------------------------------------------------------------

ALIASES: dict[str, list[str]] = {
    "tasks": ["tasknotes"],
    "scout": ["scout-beta", "scout-prod"],
    "karma": ["starlight-karma-bot-beta", "starlight-karma-bot-prod"],
}


def resolve_targets(names: list[str], *, valid_targets: set[str] | None = None) -> list[str]:
    """Expand aliases and validate target names.

    Args:
        names: User-provided target names (may include aliases).
        valid_targets: If provided, validate resolved names against this set.

    Returns:
        Deduplicated list of resolved target names.

    Raises:
        ValueError: If a resolved name is not in valid_targets.
    """
    resolved: list[str] = []
    seen: set[str] = set()
    for name in names:
        expanded = ALIASES.get(name, [name])
        for target in expanded:
            if target not in seen:
                seen.add(target)
                resolved.append(target)
    if valid_targets is not None:
        for target in resolved:
            if target not in valid_targets:
                msg = f"Unknown target {target!r}. Valid targets: {sorted(valid_targets)}"
                raise ValueError(msg)
    return resolved


# ---------------------------------------------------------------------------
# Deploy target mapping (homelab-deploy orchestration)
# ---------------------------------------------------------------------------


def _image_by_name(name: str) -> dict[str, str]:
    """Look up an image push target by its short name."""
    for img in IMAGE_PUSH_TARGETS:
        if img["name"] == name:
            return img
    for img in INFRA_PUSH_TARGETS:
        if img["name"] == name:
            return img
    msg = f"Unknown image name: {name!r}"
    raise KeyError(msg)


@dataclass(frozen=True)
class DeployTarget:
    """Describes everything needed to deploy a single logical application."""

    name: str
    images: list[dict[str, str]] = field(default_factory=list)
    charts: list[str] = field(default_factory=list)
    argo_apps: list[str] = field(default_factory=list)


def _build_deploy_targets() -> dict[str, DeployTarget]:
    """Construct the full DEPLOY_TARGETS mapping."""
    targets: dict[str, DeployTarget] = {}

    # --- Apps with custom images ---
    targets["birmel"] = DeployTarget(
        name="birmel",
        images=[_image_by_name("birmel")],
        charts=["birmel"],
        argo_apps=["birmel"],
    )
    targets["sentinel"] = DeployTarget(
        name="sentinel",
        images=[_image_by_name("sentinel")],
        charts=["sentinel"],
        argo_apps=["sentinel"],
    )
    targets["tasknotes"] = DeployTarget(
        name="tasknotes",
        images=[_image_by_name("tasknotes-server"), _image_by_name("obsidian-headless")],
        charts=["tasknotes"],
        argo_apps=["tasknotes"],
    )
    targets["scout-beta"] = DeployTarget(
        name="scout-beta",
        images=[_image_by_name("scout-for-lol")],
        charts=["scout-beta"],
        argo_apps=["scout-beta"],
    )
    targets["scout-prod"] = DeployTarget(
        name="scout-prod",
        images=[],
        charts=["scout-prod"],
        argo_apps=["scout-prod"],
    )
    targets["starlight-karma-bot-beta"] = DeployTarget(
        name="starlight-karma-bot-beta",
        images=[_image_by_name("starlight-karma-bot")],
        charts=["starlight-karma-bot-beta"],
        argo_apps=["starlight-karma-bot-beta"],
    )
    targets["starlight-karma-bot-prod"] = DeployTarget(
        name="starlight-karma-bot-prod",
        images=[],
        charts=["starlight-karma-bot-prod"],
        argo_apps=["starlight-karma-bot-prod"],
    )
    targets["pokemon"] = DeployTarget(
        name="pokemon",
        images=[_image_by_name("discord-plays-pokemon")],
        charts=["pokemon"],
        argo_apps=["pokemon"],
    )
    targets["better-skill-capped-fetcher"] = DeployTarget(
        name="better-skill-capped-fetcher",
        images=[_image_by_name("better-skill-capped-fetcher")],
        charts=["better-skill-capped-fetcher"],
        argo_apps=["better-skill-capped-fetcher"],
    )
    targets["status-page"] = DeployTarget(
        name="status-page",
        images=[_image_by_name("status-page-api")],
        charts=["status-page"],
        argo_apps=["status-page"],
    )

    # --- Infra apps with custom images ---
    targets["home"] = DeployTarget(
        name="home",
        images=[_image_by_name("homelab")],
        charts=["home"],
        argo_apps=["home"],
    )
    targets["dns-audit"] = DeployTarget(
        name="dns-audit",
        images=[_image_by_name("dns-audit")],
        charts=["dns-audit"],
        argo_apps=["dns-audit"],
    )
    # caddy-s3proxy image is used by the s3-static-sites chart
    targets["s3-static-sites"] = DeployTarget(
        name="s3-static-sites",
        images=[_image_by_name("caddy-s3proxy")],
        charts=["s3-static-sites"],
        argo_apps=["s3-static-sites"],
    )
    # dependency-summary image runs as a CronJob in the apps chart
    targets["dependency-summary"] = DeployTarget(
        name="dependency-summary",
        images=[_image_by_name("dependency-summary")],
        charts=["apps"],
        argo_apps=["apps"],
    )

    # --- Charts with no custom images ---
    chart_only = [
        "ddns",
        "apps",
        "redlib",
        "plausible",
        "cloudflare-tunnel",
        "media",
        "postal",
        "syncthing",
        "golink",
        "freshrss",
        "gickup",
        "grafana-db",
        "mcp-gateway",
        "kyverno-policies",
        "bugsink",
        "bazel-remote",
    ]
    for chart in chart_only:
        targets[chart] = DeployTarget(
            name=chart,
            images=[],
            charts=[chart],
            argo_apps=[chart],
        )

    return targets


DEPLOY_TARGETS: dict[str, DeployTarget] = _build_deploy_targets()

File: scripts/ci/src/catalog.ts
/**
 * Central registry of all build targets, deploy sites, Helm charts,
 * and deployment mappings used by the pipeline generator.
 *
 * Ported from the old Python catalog (ci.lib.catalog).
 */

// ---------------------------------------------------------------------------
// Container image push targets
// ---------------------------------------------------------------------------

export interface ImageTarget {
  name: string;
  versionKey: string;
  /** Workspace package name, if different from name (for change detection). */
  package?: string;
  /** Workspace packages needed for the image build (besides the main pkg). */
  neededPackages?: string[];
  /** Custom Dagger build function name (kebab-case). Defaults to "build-image". */
  buildFn?: string;
  /** Custom Dagger push function name (kebab-case). Defaults to "push-image". */
  pushFn?: string;
}

export const IMAGE_PUSH_TARGETS: ImageTarget[] = [
  { name: "birmel", versionKey: "shepherdjerred/birmel" },
  { name: "tasknotes-server", versionKey: "shepherdjerred/tasknotes-server" },
  {
    name: "scout-for-lol",
    versionKey: "shepherdjerred/scout-for-lol",
    buildFn: "build-scout-image",
    pushFn: "push-scout-image",
  },
  {
    name: "discord-plays-pokemon",
    versionKey: "shepherdjerred/discord-plays-pokemon",
    buildFn: "build-discord-plays-pokemon-image",
    pushFn: "push-discord-plays-pokemon-image",
  },
  {
    name: "discord-plays-mario-kart",
    versionKey: "shepherdjerred/discord-plays-mario-kart",
    buildFn: "build-discord-plays-mario-kart-image",
    pushFn: "push-discord-plays-mario-kart-image",
  },
  {
    name: "starlight-karma-bot",
    versionKey: "shepherdjerred/starlight-karma-bot",
  },
  { name: "streambot", versionKey: "shepherdjerred/streambot" },
  {
    name: "temporal-worker",
    package: "temporal",
    versionKey: "shepherdjerred/temporal-worker",
    buildFn: "build-temporal-worker-image",
    pushFn: "push-temporal-worker-image",
  },
  {
    name: "trmnl-dashboard",
    versionKey: "shepherdjerred/trmnl-dashboard",
    buildFn: "build-trmnl-dashboard-image",
    pushFn: "push-trmnl-dashboard-image",
  },
];

export const INFRA_PUSH_TARGETS: ImageTarget[] = [
  {
    name: "caddy-s3proxy",
    package: "homelab",
    versionKey: "shepherdjerred/caddy-s3proxy",
    buildFn: "build-caddy-s-3-proxy-image",
    pushFn: "push-caddy-s-3-proxy-image",
  },
  {
    name: "obsidian-headless",
    package: "homelab",
    versionKey: "shepherdjerred/obsidian-headless",
    buildFn: "build-obsidian-headless-image",
    pushFn: "push-obsidian-headless-image",
  },
  {
    name: "mcp-gateway",
    package: "homelab",
    versionKey: "shepherdjerred/mcp-gateway",
    buildFn: "build-mcp-gateway-image",
    pushFn: "push-mcp-gateway-image",
  },
  {
    name: "redlib",
    package: "homelab",
    versionKey: "shepherdjerred/redlib",
    buildFn: "build-redlib-image",
    pushFn: "push-redlib-image",
  },
];

// ---------------------------------------------------------------------------
// npm packages
// ---------------------------------------------------------------------------

export interface NpmPackage {
  name: string;
  dir: string;
}

export const NPM_PACKAGES: NpmPackage[] = [
  { name: "astro-opengraph-images", dir: "packages/astro-opengraph-images" },
  { name: "webring", dir: "packages/webring" },
  {
    name: "@shepherdjerred/helm-types",
    dir: "packages/homelab/src/helm-types",
  },
];

// ---------------------------------------------------------------------------
// Static site deploys
// ---------------------------------------------------------------------------

interface DeploySiteBase {
  bucket: string;
  name: string;
  url: string;
  buildDir: string;
  buildCmd: string;
  distDir: string;
  needsPlaywright?: boolean;
  workspaceDeps?: string;
  /**
   * Bucket-key prefixes (relative to `distDir`, trailing slash) holding
   * content-hashed/fingerprinted assets. The deploy syncs these with a 1-year
   * `immutable` Cache-Control and never `--delete`s them (a SeaweedFS lifecycle
   * rule prunes old hashes by age); everything else is synced `no-cache` +
   * `--delete`. Defaults to `["_astro/"]` (Astro's hashed output dir). Set `[]`
   * for sites with no fingerprinted assets, or a bundler-specific dir (Vite's
   * `assets/`, the scout SPA's `app/assets/`).
   */
  immutablePrefixes?: string[];
}

type DeploySiteBuildEnv =
  | {
      buildEnvVars?: string[];
      buildEnvPlaceholders?: never;
    }
  | {
      buildEnvVars?: never;
      buildEnvPlaceholders?: Readonly<Record<string, string>>;
    };

export type DeploySite = DeploySiteBase & DeploySiteBuildEnv;

export const DEPLOY_SITES: DeploySite[] = [
  {
    bucket: "sjer-red",
    name: "sjer.red",
    url: "https://sjer.red",
    buildDir: "packages/sjer.red",
    buildCmd: "bun run astro build",
    distDir: "packages/sjer.red/dist",
    needsPlaywright: true,
    workspaceDeps: "astro-opengraph-images,webring",
  },
  {
    bucket: "resume",
    name: "resume",
    url: "https://resume.sjer.red",
    buildDir: "packages/resume",
    buildCmd: "true", // pre-built by latexBuild Dagger function; deploy syncs existing files
    distDir: "packages/resume",
  },
  {
    bucket: "webring",
    name: "webring",
    url: "https://webring.sjer.red",
    buildDir: "packages/webring",
    buildCmd: "bun run typedoc",
    distDir: "packages/webring/docs",
  },
  {
    bucket: "cook",
    name: "cooklang-rich-preview",
    url: "https://cook.sjer.red",
    buildDir: "packages/cooklang-rich-preview",
    buildCmd: "bun run astro build",
    distDir: "packages/cooklang-rich-preview/dist",
  },
  {
    bucket: "stocks-sjer-red",
    name: "stocks-sjer-red",
    url: "https://stocks.sjer.red",
    buildDir: "packages/stocks-sjer-red",
    buildCmd: "bun run astro build",
    distDir: "packages/stocks-sjer-red/dist",
  },
  {
    bucket: "scout-frontend",
    name: "scout-for-lol frontend + app (prod)",
    url: "https://scout-for-lol.com",
    buildDir: "packages/scout-for-lol",
    buildCmd: "bun run scripts/build-bucket.ts",
    distDir: "packages/scout-for-lol/packages/frontend/dist",
    buildEnvVars: ["PUBLIC_PINTEREST_TAG_ID", "PUBLIC_REDDIT_PIXEL_ID"],
    workspaceDeps: "packages/frontend,packages/app",
    // Astro marketing (`_astro/`) + the Vite SPA bundle (`app/assets/`) are
    // content-hashed → immutable. The SPA shell `app/index.html` is in pass 2
    // (no-cache) so deploys take effect.
    immutablePrefixes: ["_astro/", "app/assets/"],
  },
  {
    bucket: "scout-frontend-beta",
    name: "scout-for-lol frontend + app (beta)",
    url: "https://beta.scout-for-lol.com",
    buildDir: "packages/scout-for-lol",
    buildCmd: "bun run scripts/build-bucket.ts",
    distDir: "packages/scout-for-lol/packages/frontend/dist",
    // Analytics pixels intentionally omitted for beta — beta traffic must
    // not inflate prod Pinterest/Reddit conversion data.
    buildEnvPlaceholders: {
      PUBLIC_PINTEREST_TAG_ID: "beta-placeholder-pinterest-tag-id",
      PUBLIC_REDDIT_PIXEL_ID: "beta-placeholder-reddit-pixel-id",
    },
    workspaceDeps: "packages/frontend,packages/app",
    // Astro marketing (`_astro/`) + the Vite SPA bundle (`app/assets/`) are
    // content-hashed → immutable. The SPA shell `app/index.html` is in pass 2
    // (no-cache) so deploys take effect.
    immutablePrefixes: ["_astro/", "app/assets/"],
  },
  {
    bucket: "better-skill-capped",
    name: "better-skill-capped",
    url: "https://better-skill-capped.com",
    buildDir: "packages/better-skill-capped",
    buildCmd: "bun run build",
    distDir: "packages/better-skill-capped/dist",
    // Vite SPA — content-hashed bundles live under `assets/`, not `_astro/`.
    immutablePrefixes: ["assets/"],
  },
  {
    bucket: "glitter-boys-ppl",
    name: "glitter",
    url: "https://ppl.glitter-boys.com",
    buildDir: "packages/glitter",
    buildCmd: "true",
    distDir: "packages/glitter/public",
  },
];

/** Derived from NPM_PACKAGES — workspace packages whose changes should trigger npm publishes. */
export const PACKAGES_WITH_NPM: Set<string> = new Set(
  NPM_PACKAGES.map(
    (pkg) => pkg.dir.replace("packages/", "").split("/")[0] ?? pkg.name,
  ),
);

/**
 * Top-level NPM packages that should get a per-package build step.
 * Excludes nested packages (e.g. helm-types under homelab) where the
 * workspace root doesn't have its own build script.
 */
export const NPM_BUILD_PACKAGES: Set<string> = new Set(
  NPM_PACKAGES.filter((pkg) => !pkg.dir.includes("/src/")).map((pkg) =>
    pkg.dir.replace("packages/", ""),
  ),
);

/** Maps workspace package name to the NpmPackage names it triggers. */
export const PACKAGE_TO_NPM: Record<string, string[]> = {};
for (const pkg of NPM_PACKAGES) {
  const ws = pkg.dir.replace("packages/", "").split("/")[0] ?? pkg.name;
  (PACKAGE_TO_NPM[ws] ??= []).push(pkg.name);
}

// ---------------------------------------------------------------------------
// OpenTofu stacks
// ---------------------------------------------------------------------------

export const TOFU_STACKS = [
  "cloudflare",
  "github",
  "seaweedfs",
  "tailscale",
  "buildkite",
] as const;

export const TOFU_STACK_LABELS: Record<string, string> = {
  cloudflare: "Cloudflare DNS",
  github: "GitHub Config",
  seaweedfs: "SeaweedFS Config",
  tailscale: "Tailscale ACLs",
  buildkite: "Buildkite Config",
};

// ---------------------------------------------------------------------------
// Helm charts (output by cdk8s synth)
// ---------------------------------------------------------------------------

export const HELM_CHARTS: string[] = [
  "ddns",
  "apps",
  "scout-beta",
  "scout-prod",
  "starlight-karma-bot-beta",
  "starlight-karma-bot-prod",
  "redlib",
  "plausible",
  "birmel",
  "cloudflare-tunnel",
  "media",
  "home",
  "postal",
  "syncthing",
  "golink",
  "freshrss",
  "pinchtab",
  "pokemon",
  "mario-kart",
  "gickup",
  "grafana-db",
  "mcp-gateway",
  "s3-static-sites",
  "kyverno-policies",
  "bugsink",
  "tasknotes",
  "temporal",
  "trmnl-dashboard",
];

// ---------------------------------------------------------------------------
// Package-to-site mapping (for change detection)
// ---------------------------------------------------------------------------

/**
 * Maps a workspace package name to the deploy buckets its changes should trigger.
 *
 * Most packages fan out to a single bucket. `scout-for-lol` fans out to both
 * prod and beta buckets because the merged Astro + SPA build is deployed to
 * both stages on every main merge.
 */
export const PACKAGE_TO_SITE: Record<string, string[]> = {
  "sjer.red": ["sjer-red"],
  resume: ["resume"],
  webring: ["webring"],
  "cooklang-rich-preview": ["cook"],
  "scout-for-lol": ["scout-frontend", "scout-frontend-beta"],
  "stocks-sjer-red": ["stocks-sjer-red"],
  "better-skill-capped": ["better-skill-capped"],
  glitter: ["glitter-boys-ppl"],
};

// ---------------------------------------------------------------------------
// Packages that have container image push targets
// ---------------------------------------------------------------------------

/** Derived from IMAGE_PUSH_TARGETS — packages whose changes should trigger image builds. */
export const PACKAGES_WITH_IMAGES: Set<string> = new Set(
  IMAGE_PUSH_TARGETS.map((img) => img.package ?? img.name),
);

// ---------------------------------------------------------------------------
// All packages (used when buildAll = true)
// ---------------------------------------------------------------------------

export const ALL_PACKAGES: string[] = [
  "anki",
  "astro-opengraph-images",
  "better-skill-capped",
  "birmel",
  "cooklang-for-obsidian",
  "cooklang-rich-preview",
  "discord-plays-pokemon",
  "discord-plays-mario-kart",
  "discord-stream-lifecycle",
  "discord-video-stream",
  "docs",
  "dotfiles",
  "eslint-config",
  "fonts",
  "glitter",
  "home-assistant",
  "homelab",
  "leetcode",
  "llm-models",
  "llm-observability",
  "monarch",
  "resume",
  "scout-for-lol",
  "sjer.red",
  "starlight-karma-bot",
  "stocks-sjer-red",
  "streambot",
  "tasknotes-server",
  "tasknotes-types",
  "temporal",
  "tasks-for-obsidian",
  "terraform-provider-asuswrt",
  "toolkit",
  "trmnl-dashboard",
  "webring",
];

// ---------------------------------------------------------------------------
// Resource tiers for per-package build steps
// ---------------------------------------------------------------------------

type ResourceTier = { cpu: string; memory: string };

// BK pods are thin dagger CLI wrappers — all compute happens in the remote
// Dagger engine. Keep requests minimal so more jobs fit within Kueue quota.
const HEAVY: ResourceTier = { cpu: "250m", memory: "512Mi" };
const MEDIUM: ResourceTier = { cpu: "150m", memory: "384Mi" };
const LIGHT: ResourceTier = { cpu: "100m", memory: "256Mi" };

export const PACKAGE_RESOURCES: Record<string, ResourceTier> = {
  homelab: HEAVY,
  birmel: MEDIUM,
  "scout-for-lol": MEDIUM,
  "discord-plays-pokemon": MEDIUM,
  "discord-plays-mario-kart": MEDIUM,
  // Vendored fork; its `test` loads node-av's native ffmpeg, so give it headroom.
  "discord-video-stream": MEDIUM,
  "starlight-karma-bot": MEDIUM,
  streambot: MEDIUM,
  "tasknotes-server": MEDIUM,
  "better-skill-capped": MEDIUM,
  "sjer.red": MEDIUM,
};

export { LIGHT as DEFAULT_RESOURCES };

// ---------------------------------------------------------------------------
// Packages requiring special handling
// ---------------------------------------------------------------------------

/**
 * Packages with no per-package CI steps.
 * They're still in ALL_PACKAGES for change detection (files trigger full-build awareness),
 * but perPackageSteps() skips them.
 */
export const SKIP_PACKAGES: Set<string> = new Set([
  "anki",
  "docs",
  "dotfiles",
  "fonts",
]);

/** Packages that need `bun run generate` before lint/typecheck/test (Prisma). */
export const PRISMA_PACKAGES: Set<string> = new Set([
  "birmel",
  "scout-for-lol",
  "discord-plays-mario-kart",
]);

/**
 * Packages whose runtime image needs the `gh` and `claude` CLIs preinstalled.
 *
 * Birmel's editor sub-agent shells out to both — without them the image
 * runs but the editor agent's tools no-op with a runtime warning.
 */
export const EDITOR_CLI_PACKAGES: Set<string> = new Set(["birmel"]);

/** Packages that are Astro sites. */
export const ASTRO_PACKAGES: Set<string> = new Set([
  "sjer.red",
  "cooklang-rich-preview",
]);

/** Packages that need Playwright browser tests (run in Playwright container, not bunBase). */
export const PLAYWRIGHT_PACKAGES: Set<string> = new Set(["sjer.red"]);

// ---------------------------------------------------------------------------
// Target aliases
// ---------------------------------------------------------------------------

export const ALIASES: Record<string, string[]> = {
  tasks: ["tasknotes"],
  scout: ["scout-beta", "scout-prod"],
  karma: ["starlight-karma-bot-beta", "starlight-karma-bot-prod"],
};

// ---------------------------------------------------------------------------
// Deploy targets (homelab-deploy orchestration)
// ---------------------------------------------------------------------------

export interface DeployTarget {
  name: string;
  images: ImageTarget[];
  charts: string[];
  argoApps: string[];
}

function imageByName(name: string): ImageTarget {
  const img = [...IMAGE_PUSH_TARGETS, ...INFRA_PUSH_TARGETS].find(
    (i) => i.name === name,
  );
  if (!img) {
    throw new Error(`Unknown image name: ${name}`);
  }
  return img;
}

export const DEPLOY_TARGETS: Record<string, DeployTarget> = {
  birmel: {
    name: "birmel",
    images: [imageByName("birmel")],
    charts: ["birmel"],
    argoApps: ["birmel"],
  },
  tasknotes: {
    name: "tasknotes",
    images: [imageByName("tasknotes-server")],
    charts: ["tasknotes"],
    argoApps: ["tasknotes"],
  },
  "scout-beta": {
    name: "scout-beta",
    images: [imageByName("scout-for-lol")],
    charts: ["scout-beta"],
    argoApps: ["scout-beta"],
  },
  "scout-prod": {
    name: "scout-prod",
    images: [],
    charts: ["scout-prod"],
    argoApps: ["scout-prod"],
  },
  "starlight-karma-bot-beta": {
    name: "starlight-karma-bot-beta",
    images: [imageByName("starlight-karma-bot")],
    charts: ["starlight-karma-bot-beta"],
    argoApps: ["starlight-karma-bot-beta"],
  },
  "starlight-karma-bot-prod": {
    name: "starlight-karma-bot-prod",
    images: [],
    charts: ["starlight-karma-bot-prod"],
    argoApps: ["starlight-karma-bot-prod"],
  },
  pokemon: {
    name: "pokemon",
    images: [imageByName("discord-plays-pokemon")],
    charts: ["pokemon"],
    argoApps: ["pokemon"],
  },
  media: {
    name: "media",
    images: [imageByName("streambot")],
    charts: ["media"],
    argoApps: ["media"],
  },
  "mario-kart": {
    name: "mario-kart",
    images: [imageByName("discord-plays-mario-kart")],
    charts: ["mario-kart"],
    argoApps: ["mario-kart"],
  },
  home: {
    name: "home",
    images: [],
    charts: ["home"],
    argoApps: ["home"],
  },
  "s3-static-sites": {
    name: "s3-static-sites",
    images: [imageByName("caddy-s3proxy")],
    charts: ["s3-static-sites"],
    argoApps: ["s3-static-sites"],
  },
};
